enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,502 @@
+
+1. Under-request memory resources for virtual machines by having the VM
+   pods (`launcher`) `requests.memory` being smaller than the VMs `guest.memory`
+   according to a configured ratio[^1]. This is owned by
+   [KubeVirt and present today](https://kubevirt.io/user-guide/operations/node_overcommit/).
+2. Compensate for the memory over-committment by using SWAP in order
+   to extend the virtual memory. This is owned by the platform and not
+   available today.
+
+#### Applicability
+
+We expect to mitigate the following situations
+
+* Underutilized guest memory (as described above)
+* Infrastructure related memory spikes in an over-committed environment
+* Considerable amount of "cold memory allocations" - thus allocations
+  which are rarely used
+
+#### Scope
+
+Following the upstream kubernetes approach every workload marked as burstable QoS would be able to swap.
+There is no differentiation between the type of the workload: regular pod or a VM.
+With that being said, swapping will be allowed by WASP (and later on by kube swap) for pods
+with Burstable QoS class.
+
+Among the VM workloads, VMs of high-performance configuration (NUMA affinity, CPU affinity, etc.) cannot overcommit.
+Also, VMs with best-effort QoS class don't exist because requesting memory is mandatory for the VM spec. 
+For VMs of Burstable Qos Class over-commited VM stability can be achieved during memory spikes by swapping out "cold" memory pages.
+
+#### Timeline & Phases
+
+| Phase                                                              | Target       | WASP Graduation   | Kube swap Graduation | Machine-level configuration deployment |
+|--------------------------------------------------------------------|--------------|-------------------|----------------------|----------------------------------------|
+| Phase 1 - Out-Of-Tree SWAP with WASP TP                            | mid 2024     | Tech-Preview      | Beta                 | Manual                                 |
+| Phase 2 - WASP GA with swap-based evictions                        | end 2024     | GA                | Beta                 | Manual                                 |
+| Phase 3 - Transition to Kubernetes SWAP. Limited to CNV-only users | mid-end 2025 | Start Deprecation | GA                   | Manual                                 |
+| Phase 4 - Kubernetes SWAP for all OpenShift users                  | TBD          | Removed           | GA                   | Operator                               |
+
+Because [Kubernetes SWAP] is currently in Beta and is only expected to GA within
+Kubernetes releases 1.33-1.35 (discussion about its GA criterias is still ongoing).
+this proposal is taking a four-phased approach in order to meet the timeline requirements.
+
+* **Phase 1** - OpenShift Virtualization will provide an out-of-tree
+  solution (WASP) to enable higher workload density and swap-based eviction mechanisms.
+* **Phase 2** - WASP will include swap-based eviction mechanisms.
```

Comment:
can you describe these have been done (we're in phase 2 right?)

Reply:
Done